lib/ansible/executor/play_iterator.py

```diff
@@ -400,6 +439,41 @@ def _get_next_task_from_state(self, state, host):
                             task = None
                         state.cur_always_task += 1
 
+            elif state.run_state == IteratingStates.HANDLERS:
+                if not state._handlers_sorted:
+                    # handlers are executed in the order they are defined, not in the order notified
+                    if state.handlers:
+                        all_handlers = [h for b in self._play.handlers for h in b.block]
+                        state._notified_handlers = deque((h for h in all_handlers if h in state.handlers))
+                    state._handlers_sorted = True
+
+                if state.handlers_child_state:
+                    state.handlers_child_state, task = self._get_next_task_from_state(state.handlers_child_state, host=host)
+                    if self._check_failed_state(state.handlers_child_state):
+                        state.handlers_child_state = None
+                        self._set_failed_state(state)
+                    else:
+                        if task is None or state.handlers_child_state.run_state == IteratingStates.COMPLETE:
+                            state.handlers_child_state = None
+                            continue
+                else:
+                    if state.fail_state != FailedStates.NONE and not self._play.force_handlers:
```

Comment:
`flush_handlers` in an always section only runs handlers on hosts that haven't failed.
```yaml
- hosts: host1,host2
  gather_facts: no
  handlers:
    - name: handler1
      debug:
  tasks:
    - block:
        - fail:
          when: "inventory_hostname == 'host1'"
      always:
        - command: echo
          notify: handler1

        - meta: flush_handlers
```

Would this break something?
```suggestion
                    flushed_in_always = state.pre_flushing_run_state == IteratingStates.ALWAYS and state.fail_state != FailedStates.ALWAYS
                    if state.fail_state != FailedStates.NONE and not self._play.force_handlers and not flushed_in_always:
```

Reply:
👍 

This also seems to break when `flush_handlers` is ran in the `rescue` section so we would need to amend your proposed conditional to include that case as well.

BTW note that we can't do `state.fail_state != FailedStates.ALWAYS` because there can be multiple failed states; play can fail in a block (`FailedStates.TASKS`) and then in the rescue section too (`FailedStates.RESCUE`) so we set corresponding bits in `state.fail_state` as opposed to just assign only one of `FailedStates` into it.

Reply:
Done. It appears that doing `state.pre_flushing_run_state not in (IteratingStates.RESCUE, IteratingStates.ALWAYS)` in addition to the existing conditional is enough. I think we do not have to check whether we failed in `rescue` and/or `always` because if we did that means the failure happened before `- meta: flush_handlers` and so flushing wouldn't happen. Can you think of a scenario that would need that check?

Reply:
Yeah, that makes sense. I can't think of a scenario that would need it.